Headway/BookPlayerFeature.swift

```diff
@@ -0,0 +1,215 @@
+
+            case .selectPreviousChapter:
+                guard let book = state.book else { return .none }
+                let newIndex = max(0, state.selectedChapterIndex - 1)
+                return updateChapter(state: &state, newIndex: newIndex, in: book)
+
+            case .selectNexChapter:
+                guard let book = state.book else { return .none }
+                let newIndex = min(state.selectedChapterIndex + 1, book.chapters.count - 1)
+                return updateChapter(state: &state, newIndex: newIndex, in: book)
+
+            case let .seek(progress):
+                state.playbackProgress = progress
+                return runPlayerTask { try await self.playerClient.seek(progress) }
+
+            case let .updateProgress(currentTime):
+                guard state.duration > 0 else { return .none }
+                let newProgress = currentTime / state.duration
+                if newProgress >= (1.0 - 0.01) {
+                    return .send(.chapterFinished)
+                } else {
+                    state.playbackProgress = newProgress
+                    return .none
+                }
+
+            case .chapterFinished:
+                guard let book = state.book, !book.chapters.isEmpty else { return .none }
+                let nextIndex = (state.selectedChapterIndex + 1) % book.chapters.count
+                return updateChapter(state: &state, newIndex: nextIndex, in: book)
+
+            case let .playerErrored(message):
+                state.alert = Self.makeAlert(message: message)
+                return .none
+
+            case .alert(.presented(.dismiss)):
+                state.alert = nil
+                return .none
+
+            case .alert:
+                return .none
+            }
+        }
+        .ifLet(\.$alert, action: \.alert)
+    }
+
+    private static func makeAlert(message: String) -> AlertState<Action.Alert> {
+        AlertState {
+            TextState("Something went wrong!")
+        } actions: {
+            ButtonState(role: .cancel, action: .dismiss) {
+                TextState("OK")
+            }
+        } message: {
+            TextState(message)
+        }
+    }
+}
+
+private extension BookPlayerFeature {
+    func runPlayerTask(_ task: @Sendable @escaping () async throws -> Void) -> Effect<Action> {
+        .run { send in
+            do {
+                try await task()
+            } catch {
+                await send(.playerErrored(error.localizedDescription))
+            }
+        }
+    }
+
+    func startProgressLoop() -> Effect<Action> {
+        .run { send in
+            while !Task.isCancelled {
+                try await self.clock.sleep(for: .seconds(0.5))
+                let currentTime = try await self.playerClient.currentTime()
+                await send(.updateProgress(currentTime))
+            }
+        }
+        .cancellable(id: CancelID.progress, cancelInFlight: true)
+    }
```

Comment:
Good enough approach to track progress, but, obviously, some native player mechanism would be much better and more reliable.